### PR TITLE
Expose ReadyToRun image information to profilers

### PR DESCRIPTION
**Pull request by mjsabby**
Allows profiling api consumers to know if a native image they encountered is in fact a ReadyToRun image.

### CODE REVIEW

**Comment by mjsabby** [comment]
cc @noahfalk @jkotas @erozenfeld 

This PR may be premature in that R2R images vs NGEN images are identical from a profiler's perspective. The only time (I think) a profiler even cares to know if an image maybe R2R (or plain NGEN) is when it may want to rejit a method inside one of these images.

I've followed some of @erozenfeld's discussions around R2R and it seems that the prolog nops for rejit are the same between R2R and plain NGEN images so they "should" be identical for this purpose.

There was some EH funclet discussion and thereafter change which I didn't quite get but seemed reasonable.

Let me know if there is a difference and this distinction will be useful.

File: src/vm/proftoeeinterfaceimpl.cpp [hunk 1]
```diff
@@ -4110,6 +4110,14 @@ DWORD ProfToEEInterfaceImpl::GetModuleFlags(Module * pModule)
 
         // Intentionally not checking for flat, since NGEN PEImages never have flat
         // layouts.
+
+#ifdef FEATURE_READYTORUN
```
Comment (noahfalk) [inline src/vm/proftoeeinterfaceimpl.cpp]: I want to support the migration of older profilers that may not be aware of R2R - throwing a new bit at them might be a breaking change. It feels like we need some way for profilers to proactively opt-in to these changes. 
Kshama Pawar on my team has been looking into this and I'd like to get her into this discussion, but I don't know if she has a GitHub account yet.

File: src/vm/proftoeeinterfaceimpl.cpp [hunk 2]
```diff
@@ -4110,6 +4110,14 @@ DWORD ProfToEEInterfaceImpl::GetModuleFlags(Module * pModule)
...
+        PEImageLayout * pPEImageLayout = pPEFile->GetLoaded();
+        if (pPEImageLayout != nullptr && pPEImageLayout->HasReadyToRunHeader())
```
Comment (noahfalk) [inline src/vm/proftoeeinterfaceimpl.cpp]: Having a R2R header means the image is capable of being used as R2R, but it is also possible to ignore that portion of the image and treat the image as a regular IL image. Depending on what the flag is trying to signify you might want pModule->IsReadyToRun() instead.

Comment (noahfalk) [inline src/vm/proftoeeinterfaceimpl.cpp]: R2R images return false for HasNativeImage(), so this code path would never be taken.

File: src/vm/proftoeeinterfaceimpl.cpp (hunk 1 above)
Comment (noahfalk) [inline src/vm/proftoeeinterfaceimpl.cpp]: I sent Kshama and you email offline.

**Comment by noahfalk** [comment]
Before we implement new functionality here it would be good to have an overall understanding of what you are trying to achieve and then consider what is the best way to achieve it.

File: src/vm/proftoeeinterfaceimpl.cpp (hunk 1 above)
Comment (kspawa) [inline src/vm/proftoeeinterfaceimpl.cpp]: This is Kshama and I have an account @kspawa. Will look at this at this pr tomorrow. Thanks!

**Comment by kspawa** [comment]
Agree with Noah that a broader picture will be very helpful in understanding what you are trying to achieve and the best way to achieve it.

**Comment by kspawa** [comment]
There are also considerations in terms of:
Is the R2R image actually going to be used.
When is the profiler attached with respect to the module load. If it is attached before any module loads, the R2R image could be ignored to support profiling.
What happens to the error codes we supply if the profiler is attached after module loads? Should we give a new error code corresponding to R2R images or should we send out something known to be backward compatible with existing profilers?

**Comment by mjsabby** [comment]
The ultimate goal is to understand the feasibility of rejitting a function residing in a ReadyToRun image. I would also say jitting it in the first place is also something that is interesting but now you're getting into territory that is unexposed from crossgen today (i.e. /ReadyToRun + /ProfileEnhanced if such an option existed).

The next question is if they can be rejitted through the same mechanism, i.e. we find a FunctionID somehow that belongs to an R2R image and then called RequestRejit, will other Profiler APIs that take a FunctionID work without change or not? The interesting ones are related to finding the start address of this function and its size.

**Comment by mjsabby** [comment]
@kspawa has superseded this PR by making R2R images look like NGEN images.

**Pull request closed without merging**